### ISSUE TITLE
Type narrowing fails due to unrelated instance assignment

### ISSUE THREAD

**Issue by mborgerson** [pylint-dev/pylint#10784]
### Bug description

In the below minimized case, pylint reports a false positive `no-member` error in the `check_block_local` function, following explicit object type narrowing in an expression that also accesses object attributes.

Unrelated function `make_block_with_conditional_jump` appears to influence type inference of `check_block_local`'s locally instantiated `block` object.
When `make_block_with_conditional_jump` is absent, no lint errors are generated.

In `check_block_local_separated`, separating the `block` variable type check to a separate `if` condition does pass as expected. `check_block_argument` which does not construct a `Block` locally also passes as expected.

```python
class Statement:
    pass


class Expression:
    pass


class Assignment(Statement):
    src: Expression


class ConditionalJump(Statement):
    pass


class BinaryOp(Expression):
    op: str


class Block:
    statements: list[Statement]


def make_block_with_conditional_jump() -> Block:
    cond_block = Block()
    cond_block.statements = [ConditionalJump()]
    return cond_block


def check_block_local():
    block = Block()
    for stmt in block.statements:
        if isinstance(stmt, Assignment) and isinstance(stmt.src, BinaryOp):  # E1101: Instance of 'ConditionalJump' has no 'src' member (no-member)
            print("Assignment")


def check_block_local_separated():
    block = Block()
    for stmt in block.statements:
        if isinstance(stmt, Assignment):
            if isinstance(stmt.src, BinaryOp):
                print("Assignment")


def check_block_argument(block: Block):
    for stmt in block.statements:
        if isinstance(stmt, Assignment) and isinstance(stmt.src, BinaryOp):
            print("Assignment")
```

Possibly connected to #4920

### Configuration

```ini

```

### Command used

```shell
pylint --disable=all --enable=no-member repro.py
```

### Pylint output

```python
$ pylint --disable=all --enable=no-member repro.py
************* Module repro
repro.py:34:55: E1101: Instance of 'ConditionalJump' has no 'src' member (no-member)

------------------------------------------------------------------
Your code has been rated at 8.39/10 (previous run: 8.39/10, +0.00)
```

### Expected behavior

`no-member` error should not be reported. 

### Pylint version

```shell
pylint==4.0.4
astroid==4.0.2
Python 3.12.3
```

### OS / Environment

Ubuntu 22.04

### Additional dependencies

```python

```
